logicalplan/merge_selects.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// MergeSelectsOptimizer optimizes a binary expression where
+// one select is a superset of the other select.
+// For example, the expression:
+//
+//	metric{a="b", c="d"} / scalar(metric{a="b"}) becomes:
+//	Filter(c="d", metric{a="b"}) / scalar(metric{a="b"}).
+//
+// The engine can then cache the result of `metric{a="b"}`
+// and apply an additional filter for {c="d"}.
+type MergeSelectsOptimizer struct{}
+
+func (m MergeSelectsOptimizer) Optimize(expr parser.Expr) parser.Expr {
+	heap := make(matcherHeap)
+	extractSelectors(heap, expr)
+	replaceMatchers(heap, &expr)
+
+	return expr
+}
+
+func extractSelectors(selectors matcherHeap, expr parser.Expr) {
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
+		e, ok := node.(*parser.VectorSelector)
+		if !ok {
+			return nil
+		}
+		for _, l := range e.LabelMatchers {
+			if l.Name == labels.MetricName {
+				selectors.add(l.Name, e.LabelMatchers)
+			}
+		}
+		return nil
+	})
+}
+
+func replaceMatchers(selectors matcherHeap, expr *parser.Expr) {
+	traverse(expr, func(node *parser.Expr) {
+		e, ok := (*node).(*parser.VectorSelector)
+		if !ok {
+			return
+		}
+
+		for _, l := range e.LabelMatchers {
+			if l.Name == labels.MetricName {
+				replacement, found := selectors.findReplacement(l.Name, e.LabelMatchers)
+				if found {
+					// All replacements are done on metrics only,
+					// so we can drop the explicit metric name selector.
+					filters := dropMetricName(e.LabelMatchers)
+					e.LabelMatchers = replacement
+					*node = &FilteredSelector{
+						Filters:        filters,
+						VectorSelector: e,
+					}
+					return
+				}
+			}
+		}
+	})
+}
+
+func dropMetricName(originalMatchers []*labels.Matcher) []*labels.Matcher {
+	for i, l := range originalMatchers {
+		if l.Name == labels.MetricName {
+			originalMatchers = append(originalMatchers[:i], originalMatchers[i+1:]...)
+		}
+	}
+	return originalMatchers
+}
+
+func matcherToMap(matchers []*labels.Matcher) map[string]*labels.Matcher {
+	r := make(map[string]*labels.Matcher, len(matchers))
+	for i := 0; i < len(matchers); i++ {
+		r[matchers[i].Name] = matchers[i]
+	}
+	return r
+}
+
+// matcherHeap is a set of the most selective label matchers
```

Comment:
What does it mean by selective here? Fewer postings?

Reply:
Most selective means the one that selects the most amount of postings. The idea is to select the highest amount of postings we can in a single select, and then filter down in series in selectors that don't need all postings. Let me know how we can change the wording here to make it more clear.

Reply:
Thanks, I modified the comment a bit.